### ISSUE TITLE
Increasing memory consumption with fast_executemany and varchar(max)

### ISSUE THREAD

**Issue by gordthompson** [mkleehammer/pyodbc#802]
### Environment

To diagnose, we usually need to know the following, including version numbers.  On Windows, be
sure to specify 32-bit Python or 64-bit:

- Python: 3.7 x64
- pyodbc: 4.0.30
- OS: Windows 8.1
- DB: SQL Server
- driver: ODBC Driver 17 for SQL Server

### Issue
Following up on a tangent to a SQLAlchemy issue starting [here](https://github.com/sqlalchemy/sqlalchemy/issues/5334#issuecomment-668179120) I have found that `fast_executemany` will cause a Python app to have continuously increasing memory consumption when it performs repeated `.executemany` calls involving `varchar(max)` columns. The same is not true when `varchar(100)` columns are involved.

Repro code:

``` python
from time import time
import psutil
import pyodbc

varchar_limit = "max"
print(f"using varchar({varchar_limit})")

print(f"pyodbc {pyodbc.version}")
cnxn = pyodbc.connect("DSN=mssqlLocal64", autocommit=True)
process = psutil.Process()


def print_status(msg, t0=None):
    s = f"{msg}: "
    mb = process.memory_info().vms / 1048576
    s += f"vms {mb:0.1f} MiB"
    if t0:
        s += f", {(time() - t0):0.1f} sec."
    print(s)


print_status("startup")
num_rows = 100_000
data = [(i, f"col{i:06}", 3.14159265 * (i + 1)) for i in range(num_rows)]
print_status("data loaded")

table_name = "pd_test"
col_names = ["id", "txt_col", "float_col"]
ins_sql = f"INSERT INTO {table_name} ({','.join(col_names)}) VALUES ({','.join('?' * len(col_names))})"
for iteration in range(5):
    t0 = time()
    crsr = cnxn.cursor()
    crsr.execute(f"DROP TABLE IF EXISTS {table_name}")
    crsr.execute(f"CREATE TABLE {table_name} (id int, txt_col varchar({varchar_limit}), float_col float(53))")
    crsr.fast_executemany = True
    crsr.executemany(ins_sql, data)
    crsr.close()
    print_status(f"iteration {iteration}", t0)
```

Console output:

``` none
using varchar(100)
pyodbc 4.0.30
startup: vms 8.7 MiB
data loaded: vms 31.1 MiB
iteration 0: vms 34.2 MiB, 53.8 sec.
iteration 1: vms 34.2 MiB, 47.5 sec.
iteration 2: vms 34.2 MiB, 53.6 sec.
iteration 3: vms 34.2 MiB, 50.0 sec.
iteration 4: vms 34.2 MiB, 50.8 sec.

using varchar(max)
pyodbc 4.0.30
startup: vms 8.6 MiB
data loaded: vms 30.3 MiB
iteration 0: vms 39.7 MiB, 57.1 sec.
iteration 1: vms 45.9 MiB, 58.6 sec.
iteration 2: vms 52.2 MiB, 56.9 sec.
iteration 3: vms 58.4 MiB, 56.0 sec.
iteration 4: vms 64.4 MiB, 64.7 sec.
```

**Comment by v-chojas** [comment]
Possibly related: https://github.com/mkleehammer/pyodbc/pull/703

**Comment by ScriptKiddieEvolved** [comment]
Until this bug is fixed for anyone trying to work around it, so posting it incase someone finds it useful:

The workaround does add some run time, but "hacks" around the memory management issue. Works for my use case.

Basically creating a new python process on every iteration to destroy the objects, getting around the memory issue.

Tested with multiprocessing and just recalling the function on the thread. Can clearly see the issue present when just running on the main thread vs using the workaround.


`from time import time
from multiprocessing import Process
import psutil
import pyodbc


def print_status(process,msg, time_0=None):
        s = f"{msg}: "
        mb = process.memory_info().vms / 1048576
        s += f"vms {mb:0.1f} MiB"
        if time_0:
            s += f", {(time() - time_0):0.1f} sec."
        print(s)
def function_hack(iteration,processing_type = ''):
    varchar_limit = "max"
    print(f"using varchar({varchar_limit}) with {processing_type}")

    print(f"pyodbc {pyodbc.version}")
    cnxn = pyodbc.connect(your_connection_string, autocommit=True)
    process = psutil.Process()

    print_status(process,"startup")
    num_rows = 100000
    data = [(i, f"col{i:06}", 3.14159265 * (i + 1)) for i in range(num_rows)]
    print_status(process,"data loaded")

    table_name = "pd_test"
    col_names = ["id", "txt_col", "float_col"]
    ins_sql = f"INSERT INTO {table_name} ({','.join(col_names)}) VALUES ({','.join('?' * len(col_names))})"
 
    time_0 = time()
    crsr = cnxn.cursor()
    crsr.execute(f"DROP TABLE IF EXISTS {table_name}")
    crsr.execute(f"CREATE TABLE {table_name} (id int, txt_col varchar({varchar_limit}), float_col float(53))")
    crsr.fast_executemany = True
    crsr.executemany(ins_sql, data)
    crsr.close()
    print_status(process,f"iteration {iteration}", time_0)

if __name__ == '__main__':

    for iteration in range(5):
        process_multi = Process(target = function_hack,args=(iteration,'multiprocessing',))
        process_multi.start()
        process_multi.join()
    
    for iteration in range(5):
        function_hack(iteration,'main thread')
 
`

`using varchar(max) with multiprocessing
pyodbc 4.0.30
startup: vms 13.6 MiB
data loaded: vms 33.1 MiB
iteration 0: vms 42.4 MiB, 25.1 sec.
using varchar(max) with multiprocessing
pyodbc 4.0.30
startup: vms 13.7 MiB
data loaded: vms 33.6 MiB
iteration 1: vms 42.9 MiB, 23.7 sec.
using varchar(max) with multiprocessing
pyodbc 4.0.30
startup: vms 13.6 MiB
data loaded: vms 33.1 MiB
iteration 2: vms 42.4 MiB, 26.4 sec.
using varchar(max) with multiprocessing
pyodbc 4.0.30
startup: vms 13.6 MiB
data loaded: vms 34.3 MiB
iteration 3: vms 43.6 MiB, 25.7 sec.
using varchar(max) with multiprocessing
pyodbc 4.0.30
startup: vms 13.6 MiB
data loaded: vms 33.2 MiB
iteration 4: vms 42.5 MiB, 24.7 sec.


using varchar(max) with main thread
pyodbc 4.0.30
startup: vms 13.4 MiB
data loaded: vms 33.0 MiB
iteration 0: vms 42.5 MiB, 24.7 sec.
using varchar(max) with main thread
pyodbc 4.0.30
startup: vms 20.4 MiB
data loaded: vms 39.3 MiB
iteration 1: vms 48.5 MiB, 23.9 sec.
using varchar(max) with main thread
pyodbc 4.0.30
startup: vms 27.0 MiB
data loaded: vms 45.5 MiB
iteration 2: vms 54.8 MiB, 25.0 sec.
using varchar(max) with main thread
pyodbc 4.0.30
startup: vms 33.7 MiB
data loaded: vms 51.8 MiB
iteration 3: vms 61.0 MiB, 23.9 sec.
using varchar(max) with main thread
pyodbc 4.0.30
startup: vms 40.5 MiB
data loaded: vms 58.0 MiB
iteration 4: vms 67.3 MiB, 27.3 sec.
.`

**Comment by gordthompson** [comment]
@mkleehammer - I just tested #832 and it does seem to fix this issue.

**Comment by gordthompson** [comment]
Fixed in 716572a1ca4426483bb31d39802b543d44ac6732
